frontend/src/Components/ScaleSelector/ScaleSelector.tsx

```diff
@@ -1,25 +1,128 @@
+import { useEffect, useRef, useState } from 'preact/hooks';
 import { ScaleName, SCALES } from '../../../../shared/scales';
 import { SELECT_CHANGE_SCALE } from '../../constants';
+import sharedClasses from '../../styles.module.css';
 import { connectToWebSocket } from '../WebSocket/WebSocket';
 import classes from './ScaleSelector.module.css';
 
-export const ScaleSelector = connectToWebSocket(({ socket: { connected, setScale } }) => (
-  <select
-    disabled={!connected}
-    name="scale"
-    class={classes.select}
-    onChange={({ target }) =>
-      setScale(SCALES[(target as HTMLSelectElement).value as ScaleName].values)
+const availableScales = Object.entries(SCALES);
+
+export const ScaleSelector = connectToWebSocket(({ socket: { connected, setScale } }) => {
+  const [open, setOpen] = useState(false);
+  const [selected, setSelected] = useState(-1);
+  const selectionButtonRef = useRef<HTMLButtonElement>(null);
+  const dropdownRef = useRef<HTMLUListElement>(null);
+
+  const close = () => {
+    setOpen(false);
+    setSelected(-1);
+  };
+
+  const toggle = () => {
+    if (open) {
+      close();
+    } else {
+      setSelected(0);
+      setOpen(true);
+    }
+  };
+
+  const handleBlur = ({ relatedTarget }: FocusEvent) => {
+    if (relatedTarget !== selectionButtonRef.current && relatedTarget !== dropdownRef.current) {
+      close();
+    }
+  };
+
+  const handleKeyboardNavigation = (event: KeyboardEvent) => {
+    switch (event.key) {
+      case 'ArrowDown':
+        setSelected((selected + 1) % availableScales.length);
+        event.preventDefault();
+        break;
+      case 'ArrowUp':
+        setSelected((Math.max(selected, 0) + availableScales.length - 1) % availableScales.length);
+        event.preventDefault();
+        break;
+      case 'Home':
+      case 'PageUp':
+        setSelected(0);
+        break;
+      case 'End':
+      case 'PageDown':
+        setSelected(availableScales.length - 1);
+        break;
+      case 'Escape':
+        close();
+        selectionButtonRef.current?.focus();
+        break;
+      case 'Enter':
+      case ' ':
+        // Prevent "enter" from triggering the button if the dropdown is open
+        event.preventDefault();
+        close();
+        selectionButtonRef.current?.focus();
+        if (selected >= 0) {
+          setScale(availableScales[selected][1].values);
+        }
+        break;
+    }
+  };
+
+  useEffect(() => {
+    if (!open) {
+      return;
     }
-    value={'CHANGE_SCALE'}
-  >
-    <option value="CHANGE_SCALE" selected disabled hidden>
-      {SELECT_CHANGE_SCALE}
-    </option>
-    {Object.entries(SCALES).map(([id, { name }]) => (
-      <option value={id} key={id}>
-        {name}
-      </option>
-    ))}
-  </select>
-));
+    if (!connected) {
+      close();
+      return;
+    }
+    if (dropdownRef.current) {
+      dropdownRef.current.scrollIntoView({ behavior: 'smooth', block: 'nearest' });
```

Comment:
This scroll effect is acceptable, but I would prefer if the drop down would open either to the top or bottom depending on the available space. I would assume that this much more complicated. Did you consider it?

Reply:
I considered it but there is a lot of effort involved as we do not know a priori how big the dropdown is (well, technically we can calculate its size), and as far as I know you can only do it via JavaScript. The question is also if it should switch alignment when the user scrolls or resizes the window. I will think about this longer.

Reply:
Then let's go with the current solution. We can always improve on it in a second iteration. This is definitively good enough.

Reply:
Too late. In the end, it was actually not too involved except that I make an assumption about the height of the dropdown (that statically takes the number of scales into account). This should be fine as long as we do not mess with the styling. I decided against updating the dropdown position when scrolling as that was actually only relevant when the imprint is visible, which should not be an important situation.

Reply:
Oh, that is a nice surprise! It improves handling of on mobile quite a bit.